Mathlib/Geometry/Manifold/Bordisms.lean

```diff
@@ -0,0 +1,202 @@
+  [TopologicalSpace H'] [TopologicalSpace H''] [TopologicalSpace H''']
+
+variable {M' : Type*} [TopologicalSpace M'] [ChartedSpace H' M']
+  {I' : ModelWithCorners ℝ E' H'} [IsManifold I' k M']
+  [BoundarylessManifold I' M'] [CompactSpace M'] [FiniteDimensional ℝ E']
+
+variable (M I) in
+/-- If `M` is `n`-dimensional and closed, it is a singular `n`-manifold over itself.-/
+noncomputable def refl (hdim : finrank ℝ E = n) :
+    SingularNManifold M n k I where
+  dimension := hdim
+  f := id
+  hf := continuous_id
+
+/-- If `(N, f)` is a singular `n`-manifold on `X` and `M` another `n`-dimensional manifold,
+a smooth map `φ : M → N` induces a singular `n`-manifold structure `(M, f ∘ φ)` on `X`. -/
+noncomputable def comap [h : Fact (finrank ℝ E' = n)]
+    (s : SingularNManifold X n k I)
+    {φ : M' → s.M} (hφ : ContMDiff I' I n φ) : SingularNManifold X n k I' where
+  f := s.f ∘ φ
+  hf := s.hf.comp hφ.continuous
+  dimension := h.out
+
+@[simp]
+lemma comap_f [Fact (finrank ℝ E' = n)]
+    (s : SingularNManifold X n k I) {φ : M' → s.M} (hφ : ContMDiff I' I n φ) :
+    (s.comap hφ).f = s.f ∘ φ :=
+  rfl
+
+variable (M I) in
+/-- The canonical singular `n`-manifold associated to the empty set (seen as an `n`-dimensional
+manifold, i.e. modelled on an `n`-dimensional space). -/
+def empty [h: Fact (finrank ℝ E = n)] (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+    {I : ModelWithCorners ℝ E H} [IsManifold I k M] [IsEmpty M] : SingularNManifold X n k I where
+  M := M
+  dimension := h.out
+  f := fun x ↦ (IsEmpty.false x).elim
+  hf := by
+    rw [continuous_iff_continuousAt]
+    exact fun x ↦ (IsEmpty.false x).elim
+
+variable (M I) in
+/-- An `n`-dimensional manifold induces a singular `n`-manifold on the one-point space. -/
+def trivial [h: Fact (finrank ℝ E = n)] : SingularNManifold PUnit n k I where
+  M := M
+  dimension := h.out
+  f := fun _ ↦ PUnit.unit
+  hf := continuous_const
+
+/-- The product of a singular `n`- and a singular `m`-manifold into a one-point space
+is a singular `n+m`-manifold. -/
+-- FUTURE: prove that this observation induces a commutative ring structure
+-- on the unoriented bordism group `Ω_n^O = Ω_n^O(pt)`.
+def prod {m n : ℕ} (s : SingularNManifold PUnit n k I) (t : SingularNManifold PUnit m k I') :
+    SingularNManifold PUnit (n + m) k (I.prod I') where
+  M := s.M × t.M
+  f := fun _ ↦ PUnit.unit
+  hf := continuous_const
+  dimension := by rw [finrank_prod, s.dimension, t.dimension]
+
+variable (s t : SingularNManifold X n k I)
+
+/-- The disjoint union of two singular `n`-manifolds on `X` is a singular `n`-manifold on `X`. -/
+-- We need to choose a model space for the disjoint union (as a priori `s` and `t` could be
+-- modelled on very different spaces: for simplicity, we choose `ℝ^n`; all real work is contained
+-- in the two instances above.
+def sum {n : ℕ} (s t : SingularNManifold X n k I) [finrank: Fact (finrank ℝ E = n)] :
+    SingularNManifold X n k I where
```

Comment:
I am confused by the design and the comment here. One you fix `I`, you are fixing the model space, so by definition `s` and `t` have the same model space, and there is nothing special to do: just use `I` also for the disjoint union. On the other hand, it makes me realize that the parameter `n` is completely useless in all the story, as it is also determined by `I` (it's the dimension of the target space of `I`). So why not just remove it completely from the definition of `SingularNManifold` (and from the whole file)?

Reply:
Good point. In the latest version, `n` is removed from the discussion, and only `I` is a type parameter.